Frameworks/CoreGraphics/CGContextCairo.mm

```diff
@@ -448,54 +452,78 @@
-    CGSize size;
+void CGContextCairo::CGContextShowGlyphsAtPoint(float x, float y, const CGGlyph* glyphs, size_t count) {
+    CGSize size = CGSizeZero;
 
     _isDirty = true;
 
+    std::vector<int> designUnitAdvances(count);
+    CGFontRef font = curState->getCurFont();
+    CGFontGetGlyphAdvances(font, glyphs, count, designUnitAdvances.data());
+    std::vector<CGSize> advances(count);
+    std::transform(designUnitAdvances.cbegin(),
+                   designUnitAdvances.cend(),
+                   advances.begin(),
+                   [ scale = curState->fontSize / CGFontGetUnitsPerEm(font), &size ](int unscaledAdvance) {
+                       CGFloat advanceWidth = scale * unscaledAdvance;
+                       size.width += advanceWidth;
+                       return CGSize{ advanceWidth, 0 };
+                   });
     switch (curState->textDrawingMode) {
         case kCGTextFill:
         case kCGTextStroke:
         case kCGTextFillStroke:
         case kCGTextFillClip:
         case kCGTextStrokeClip:
         case kCGTextFillStrokeClip:
-            size = CGFontDrawGlyphsToContext(glyphs, count, x, y);
+            curState->curTextPosition = { x, y };
+            CGContextShowGlyphsWithAdvances(glyphs, advances.data(), count);
             break;
 
         case kCGTextClip:
         case kCGTextInvisible:
-            // TODO #924: Update the text position in this case
-            UNIMPLEMENTED();
+            // Do nothing, set text position at end
             break;
     }
 
-    curState->curTextPosition.x = x + size.width;
-    curState->curTextPosition.y = y;
+    curState->curTextPosition = { x + size.width, y };
 }
 
-void CGContextCairo::CGContextShowGlyphsWithAdvances(WORD* glyphs, CGSize* advances, int count) {
-    CGPoint curPos = { curState->curTextPosition.x, curState->curTextPosition.y };
+void CGContextCairo::CGContextShowGlyphsWithAdvances(const CGGlyph* glyphs, const CGSize* advances, size_t count) {
     _isDirty = true;
-
-    for (int i = 0; i < count; i++) {
-        CGFontDrawGlyphsToContext(&glyphs[i], 1, curPos.x, curPos.y);
-        curPos.x += advances[i].width;
-        curPos.y += advances[i].height;
+    CGPoint curPos = curState->curTextPosition;
+    CGFontRef font = curState->getCurFont();
+    ComPtr<IDWriteFontFace> fontFace;
+    if (SUCCEEDED(_CGFontGetDWriteFontFace(font, &fontFace))) {
+        std::vector<DWRITE_GLYPH_OFFSET> positions(count);
+        CGPoint delta = CGPointZero;
+        std::transform(advances, advances + count, positions.begin(), [&delta](CGSize size) {
+            DWRITE_GLYPH_OFFSET ret = { delta.x, delta.y };
+            delta.x += size.width;
+            delta.y += size.height;
+            return ret;
+        });
+        std::vector<FLOAT> dwriteAdvances(count);
+
+        DWRITE_GLYPH_RUN run = { fontFace.Get(), curState->fontSize, count, glyphs, dwriteAdvances.data(), positions.data(), FALSE, 0 };
```

Comment:
>dwriteAdvances.data() [](start = 84, length = 21)

can you check if we can get away with leaving this as nullptr? would be great if that were possible... otherwise leave a comment about it.